buildengine/engine.go

```diff
@@ -185,6 +185,48 @@ func (e *Engine) watchForModuleChanges(ctx context.Context, period time.Duration
 	defer watch.Close()
 	watch.Subscribe(events)
 
+	moduleHashes := map[string][]byte{}
+	e.controllerSchema.Range(func(name string, sch *schema.Module) bool {
+		hash, err := computeModuleHash(sch)
+		if err != nil {
+			logger.Errorf(err, "compute hash for %s failed", name)
+			return false
+		}
+		moduleHashes[name] = hash
+		return true
+	})
+
+	// Watch for schema changes from the FTL controller.
+	go rpc.RetryStreamingServerStream(ctx, backoff.Backoff{Max: time.Second}, &ftlv1.PullSchemaRequest{}, e.client.PullSchema, func(ctx context.Context, msg *ftlv1.PullSchemaResponse) error {
```

Comment:
We're already pulling from the schema, why are we doing it twice instead of reusing the existing function?

Reply:
No strong reason other than it seemed like the code was a bit more self contained in this function. But I can def refactor to reuse the existing one. I'll need to come up with a way to start getting updates after I do the deploy, but I'll see what works there.

Reply:
For notifications I'd probably use a `pubsub.Topic`

Reply:
Sweet! That's what I'm updating to as we speak, update incoming :)